src/deterministicArchive.js

```diff
@@ -0,0 +1,129 @@
+ *
+ * @param {string} dirOrFile
+ * @returns {Promise<Array<{name: string, path: string}>>}
+ */
+async function resolveFilesRecursiveForDir(dirOrFile) {
+  const isDir = (await fs.promises.lstat(dirOrFile)).isDirectory();
+
+  if (isDir) {
+    const files = await fs.promises.readdir(dirOrFile, {
+      withFileTypes: true,
+      recursive: true,
+    });
+
+    return files
+      .filter((dirent) => dirent.isFile())
+      .map((dirent) => ({
+        name: dirent.name,
+        path: path.resolve(dirent.path, dirent.name),
+      }));
+  }
+
+  return [
+    {
+      name: path.basename(dirOrFile),
+      path: path.resolve(dirOrFile),
+    },
+  ];
+}
+
+/**
+ * Resolves all files in all directories recursively
+ *
+ * @param {...string} dirsAndFiles
+ * @returns {Promise<Array<{name: string, path: string}>>}
+ */
+async function resolveFilesRecursive(...dirsAndFiles) {
+  const files = await Promise.all(
+    dirsAndFiles.filter(Boolean).map(resolveFilesRecursiveForDir),
+  );
+
+  return files.flat();
+}
+
+/**
+ * Creates a deterministic archive of the given files
+ *
+ * @param {string[]} dirsAndFiles
+ * @param {{name: string, content: string}[]} contentToArchive
+ * @returns {Promise<{buffer: Buffer, hash: string}>}
+ */
+export default async function deterministicArchive(
+  dirsAndFiles,
+  contentToArchive = [],
+) {
+  // Sort by path to make the output deterministic
+  const filesToArchiveSorted = (await resolveFilesRecursive(...dirsAndFiles)).sort(
+    (a, b) => a.path.localeCompare(b.path),
+  );
+  const contentToArchiveSorted = contentToArchive.sort((a, b) =>
+    a.name.localeCompare(b.name),
+  );
+
+  return new Promise((resolve, reject) => {
+    const archive = new Archiver('zip', {
+      // Concurrency in the stat queue leads to non-deterministic output.
+      // https://github.com/archiverjs/node-archiver/issues/383#issuecomment-2253139948
+      statConcurrency: 1,
+      zlib: { level: 6 },
+    });
+
+    const stream = new Writable();
+    const data = [];
+
+    stream._write = (chunk, _enc, done) => {
+      data.push(...chunk);
+      done();
+    };
+
+    const entries = [];
+    archive.on('entry', (entry) => {
+      entries.push(entry);
+    });
+
+    stream.on('finish', () => {
+      validateArchive(archive.pointer(), entries);
+      const buffer = Buffer.from(data);
+      const hash = crypto.createHash('md5').update(buffer).digest('hex');
+
+      resolve({ buffer, hash });
+    });
+    archive.pipe(stream);
+
+    // We can't use archive.directory() here because it is not deterministic.
+    // https://github.com/archiverjs/node-archiver/issues/383#issuecomment-2252938075
+    for (const file of filesToArchiveSorted) {
+      archive.file(file.path, {
+        name: file.name,
+        prefix: '',
+        date: FILE_CREATION_DATE,
+      });
+    }
+
+    for (const file of contentToArchiveSorted) {
+      archive.append(file.content, {
+        name: file.name,
+        prefix: '',
```

Comment:
We didn't have prefix when using archive.append before, but since we are using it with archive.file I figured we might as well add it here for consistency.